cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/NamespaceListService.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.master.environment.k8s;
+
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import io.cdap.cdap.common.namespace.NamespaceQueryAdmin;
+import io.cdap.cdap.master.environment.MasterEnvironments;
+import io.cdap.cdap.master.spi.environment.MasterEnvironment;
+import io.cdap.cdap.master.spi.namespace.NamespaceDetail;
+import io.cdap.cdap.master.spi.namespace.NamespaceLister;
+import org.apache.twill.api.TwillRunnerService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * A service for sending the list of namespaces to the {@link TwillRunnerService}.
+ */
+public class NamespaceListService extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(NamespaceListService.class);
+
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+
+  NamespaceListService(NamespaceQueryAdmin namespaceQueryAdmin) {
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+  }
+
+  @Override
+  protected void run() throws Exception {
+    MasterEnvironment masterEnv = MasterEnvironments.getMasterEnvironment();
```

Comment:
This class shouldn't know about MasterEnvironment, as depending on the configuration, AppFabric doesn't always use the one from MasterEnvironment. Instead, inject both the NamespaceQueryAdmin and the TwillRunnerService in the constructor.

Reply:
fixed